File: data/gpt-4o/readable_format_HPEC/cpp/file149.cpp
#include "tinfo.h"
#include "cg_data.h"
#include <algorithm>  

void makea(int n, int nz, double* a, int* colidx, int* rowstr, int firstrow, int lastrow, int firstcol, int lastcol, int* arow, int* acol, double* aelt, double* v, int *iv)
{
    int i, iouter, ivelt, nzv, nn1;
    int *ivc = new int [nonzer+1];
    double *vc = new double [nonzer+1];

    nn1 = 1;
    while (nn1 < n)
        nn1 *= 2;

    num_threads = 1;
    myid = 0;
    if (num_threads > max_threads) 
    {
        if (myid == 0) 
            printf("Warning: num_threads %d exceeded an internal limit %d
", num_threads, max_threads);
        num_threads = max_threads;
    }
    int work = (n + num_threads - 1) / num_threads;
    int ilow = work * myid + 1;
    int ihigh = ilow + work - 1;
    if (ihigh > n) 
        ihigh = n;

    for (iouter = 1; iouter <= ihigh; iouter++) 
    {
        nzv = nonzer;
        sprnvc(n, nzv, nn1, vc, ivc);
        if (iouter >= ilow) 
        {
            vecset(n, vc, ivc, nzv, iouter, .5);
            arow[iouter-1] = nzv;
            for (ivelt = 0; ivelt < nzv; ivelt++) 
            {
                acol[ivelt + (iouter - 1) * (nonzer+1)] = ivc[ivelt];
                aelt[ivelt + (iouter - 1) * (nonzer+1)] = vc[ivelt];
            }
        }
    }
    sparse(a, colidx, rowstr, n, nz, nonzer, arow, acol, aelt, firstrow, lastrow, v, &iv[0], &iv[nz], rcond, shift);

    delete [] ivc;
    delete [] vc;
}

File: data/CodeLlama/readable_format_HPEC/fortran/file149.f90
subroutine makea( n, nz, a, colidx, rowstr,  &
&                  firstrow, lastrow, firstcol, lastcol,  &
&                  arow, acol, aelt, v, iv )

use tinfo
use cg_data, only : nonzer, rcond, shift

implicit none

integer             n
integer(kz)         nz, rowstr(n+1)
integer             firstrow, lastrow, firstcol, lastcol
integer             colidx(nz)
integer             iv(n+nz), arow(n), acol(nonzer+1,n)
double precision    aelt(nonzer+1,n), v(nz)
double precision    a(nz)


integer          i, iouter, ivelt, nzv, nn1
integer          ivc(nonzer+1)
double precision vc(nonzer+1)


external          sparse, sprnvc, vecset
integer           work



nn1 = 1
50   continue
nn1 = 2 * nn1
if (nn1 .lt. n) goto 50

num_threads = 1
myid = 0
if (num_threads .gt. max_threads) then
if (myid .eq. 0) write(*,100) num_threads, max_threads
100      format(' Warning: num_threads',i6,  &
&          ' exceeded an internal limit',i6)
num_threads = max_threads
endif
work  = (n + num_threads - 1)/num_threads
ilow  = work * myid + 1
ihigh = ilow + work - 1
if (ihigh .gt. n) ihigh = n

do iouter = 1, ihigh
nzv = nonzer
call sprnvc( n, nzv, nn1, vc, ivc )
if ( iouter .ge. ilow ) then
call vecset( n, vc, ivc, nzv, iouter, .5D0 )
arow(iouter) = nzv
do ivelt = 1, nzv
acol(ivelt, iouter) = ivc(ivelt)
aelt(ivelt, iouter) = vc(ivelt)
enddo
endif
enddo

call sparse( a, colidx, rowstr, n, nz, nonzer, arow, acol,  &
&             aelt, firstrow, lastrow,  &
&             v, iv(1), iv(nz+1), rcond, shift )
return

end
